shap/shap
# Added support for Independent TreeSHAP

- Wrote a linear time implementation of TreeSHAP that computes Shapley values with **independence** between the conditioned and non-conditioned sets.  
    - This implementation works for a single reference, a single sample, and a single tree.  
    - The overall runtime is O(NLT), where N is the size of the reference set, L is the number of leaves in the tree, and T is the number of trees.
    - Wrote a c++ implementation which works approximately 100 times faster than the python implementation.
    - Added support for non-linear functions of the output, in order to support decomposing loss functions to particular features.
- Added an implementation for xgboost booster models.  In addition, wrote up an implementation to create tree.py Tree objects from the string dump of the booster models.
- Added tests to test out independent TreeSHAP
    - Verify that Independent TreeSHAP gives the same results as Kernel SHAP for simple models (which makes the same independence assumptions).
    - Verify that the output sums up correctly for models with 20 trees of max_depth 6.
    - Verify that the output sums up correctly for computing the attributions of mse, logistic, and logloss.
